cmd/attest.go

```diff
@@ -0,0 +1,173 @@
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"strconv"
+
+	"cloud.google.com/go/compute/metadata"
+	"github.com/google/go-tpm-tools/client"
+	"github.com/google/go-tpm-tools/proto/attest"
+	"github.com/google/go-tpm/tpm2"
+	"github.com/spf13/cobra"
+	"google.golang.org/protobuf/proto"
+)
+
+var (
+	key      string
+	teeNonce []byte
+)
+
+var attestationKeys = map[string]map[tpm2.Algorithm]func(rw io.ReadWriter) (*client.Key, error){
+	"AK": {
+		tpm2.AlgRSA: client.AttestationKeyRSA,
+		tpm2.AlgECC: client.AttestationKeyECC,
+	},
+	"gceAK": {
+		tpm2.AlgRSA: client.GceAttestationKeyRSA,
+		tpm2.AlgECC: client.GceAttestationKeyECC,
+	},
+}
+
+var attestCmd = &cobra.Command{
+	Use:   "attest",
+	Short: "Create a remote attestation report",
+	Long: `Gather information for remote attestation.
+The Attestation report contains a quote on all available PCR banks, a way to validate 
+the quote, and a TCG Event Log (Linux only).
+Use --key to specify the type of attestation key. It can be gceAK for GCE attestation
+key or AK for a custom attestation key. By default it uses AK.
+--algo flag override the public key algorithm for attestation key. If not provided then
+by default rsa is used.
+--nonce attaches even length extra data to the report and guarantees a fresh quote.
```

Comment:
even length? I thought TPM 2.0 used 48 bytes. The "fresh quote" claim is only a guarantee if the user is providing a true nonce. It's not a claim that this library can make on its own.

Reply:
Yes, but if we give an odd length nonce , then it gives an error stating `"Error: invalid argument "12345" for "--nonce" flag: encoding/hex: odd length hex string"` .
Because it try to do hex-encoding while taking input in --nonce flag.
```
func addNonceFlag(cmd *cobra.Command) {
	cmd.PersistentFlags().BytesHexVar(&nonce, "nonce", []byte{}, "hex encoded nonce for vTPM attestation, cannot be empty")
}
```

Reply:
Nonce does not need to be 48 bytes long.
Do you want me to implement a zero filling behavior if nonce is of odd length?

Reply:
The answer here is "it depends". Specifically on the size of TPM2B_DATA, which is TPM2 implementation specific, since it also depends on the max name size. The lower bound IIRC is a 2byte hash identifier and the size of the digest for the hash function. Since TPM2s are required to support SHA256, we get a lower bound of 2+(256/8)=34. However, many TPMs support larger hash algorithms, like SHA512 which would yield 66 bytes. You can get the size by calling TPM2_GetCapability and checking TPM2_PT_MAX_DIGEST

This discussion should be moot though since we removed the nonce help text in attest in favor of the default flag help text.